### PR TITLE
enhancement(remap transform): add `parse_aws_alb_log` function

### LINKED ISSUE #5365: New `parse_aws_elb_log` Remap function
A remap function to parse AWS ELB logs. This is necessary when using the new `aws_s3` source. The logs coming out of that source are in raw format and will need parsing.

## Example

Given the following ELB logs ([more examples](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-access-logs.html#access-log-entry-examples):

```
http 2018-07-02T22:23:00.186641Z app/my-loadbalancer/50dc6c495c0c9188 
192.168.131.39:2817 10.0.0.1:80 0.000 0.001 0.000 200 200 34 366 
"GET http://www.example.com:80/ HTTP/1.1" "curl/7.46.0" - - 
arn:aws:elasticloadbalancing:us-east-2:123456789012:targetgroup/my-targets/73e2d6bc24d8a067
"Root=1-58337262-36d228ad5d99923122bbe354" "-" "-" 
0 2018-07-02T22:22:48.364000Z "forward" "-" "-" 10.0.0.1:80 200 "-" "-"
```

A user should be able to parse this with:

```js
. = parse_aws_elb(.message)
```

This would result in:

```js
{
	"type": "HTTP",
	"timestamp": "2018-07-02T22:23:00.186641Z", // should be a native timestamp
	"elb": "app/my-loadbalancer/50dc6c495c0c9188",
	"client_host": "192.168.131.39:2817",
	"target_host": "10.0.0.1:80",
	"request_processing_time": 0.000,
	"target_processing_time": 0.001,
	"response_processing_time": 0.000,
	"elb_status_code": 200
	"target_status_code": 200,
	"received_bytes": 34,
	"sent_bytes": 366,
	"request_method": "GET",
	"request_url": "http://www.example.com:80/",
	"request_protocol": "HTTP/1.1",
	"user_agent": "curl/7.46.0",
	"ssl_cipher": null,
	"ssl_protocol": null,
	"target_group_arn": "arn:aws:elasticloadbalancing:us-east-2:123456789012:targetgroup/my-targets/73e2d6bc24d8a067",
	"trace_id": "Root=1-58337262-36d228ad5d99923122bbe354",
	"domain_name": null,
	"chosen_cert_arn": null,
	"matched_rule_priority": 0,
	"request_creation_time": "2018-07-02T22:22:48.364000Z", // should be a native timestamp
	"actions_executed": "forward",
	"redirect_url": null,
	"error_reason": null
}
```

### Regex

@jszwedko wrote this in our docs that might be helpful:

To parse AWS load balancer logs, the `regex_parser` transform can be used:

```toml
[transforms.elasticloadbalancing_fields_parsed]
  type = "regex_parser"
  inputs = ["s3"]
  regex = '(?x)^
      (?P<type>[\\w]+)[ ]
      (?P<timestamp>[\\w:.-]+)[ ]
      (?P<elb>[^\\s]+)[ ]
      (?P<client_host>[\\d.:-]+)[ ]
      (?P<target_host>[\\d.:-]+)[ ]
      (?P<request_processing_time>[\\d.-]+)[ ]
      (?P<target_processing_time>[\\d.-]+)[ ]
      (?P<response_processing_time>[\\d.-]+)[ ]
      (?P<elb_status_code>[\\d-]+)[ ]
      (?P<target_status_code>[\\d-]+)[ ]
      (?P<received_bytes>[\\d-]+)[ ]
      (?P<sent_bytes>[\\d-]+)[ ]
      "(?P<request_method>[\\w-]+)[ ]
      (?P<request_url>[^\\s]+)[ ]
      (?P<request_protocol>[^"\\s]+)"[ ]
      "(?P<user_agent>[^"]+)"[ ]
      (?P<ssl_cipher>[^\\s]+)[ ]
      (?P<ssl_protocol>[^\\s]+)[ ]
      (?P<target_group_arn>[\\w.:/-]+)[ ]
      "(?P<trace_id>[^\\s"]+)"[ ]
      "(?P<domain_name>[^\\s"]+)"[ ]
      "(?P<chosen_cert_arn>[\\w:./-]+)"[ ]
      (?P<matched_rule_priority>[\\d-]+)[ ]
      (?P<request_creation_time>[\\w.:-]+)[ ]
      "(?P<actions_executed>[\\w,-]+)"[ ]
      "(?P<redirect_url>[^"]+)"[ ]
      "(?P<error_reason>[^"]+)"'
  field = "message"
  drop_failed = false

  types.received_bytes = "int"
  types.request_processing_time = "float"
  types.sent_bytes = "int"
  types.target_processing_time = "float"
  types.response_processing_time = "float"

[transforms.elasticloadbalancing_url_parsed]
  type = "regex_parser"
  inputs = ["elasticloadbalancing_fields_parsed"]
  regex = '^(?P<url_scheme>[\\w]+)://(?P<url_hostname>[^\\s:/?#]+)(?::(?P<request_port>[\\d-]+))?-?(?:/(?P<url_path>[^\\s?#]*))?(?P<request_url_query>\\?[^\\s#]+)?'
  field = "request_url"
  drop_failed = false
```

### PR DESCRIPTION
**Pull request by fanatid**
Closes #5365 

I also was curious how [nom](https://crates.io/crates/nom) will be faster than `regex`, so create a simple benchmark:

<details>
<summary>code diff</summary>

```diff
diff --git a/Cargo.toml b/Cargo.toml
index 956583b23..63dcccf4e 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -565,5 +565,9 @@ name = "remap"
 harness = false
 required-features = ["remap-benches"]
 
+[[bench]]
+name = "parse_aws_elb"
+harness = false
+
 [patch.'https://github.com/tower-rs/tower']
 tower-layer = "0.3"
diff --git a/src/remap/function.rs b/src/remap/function.rs
index 1121a922c..537b19b35 100644
--- a/src/remap/function.rs
+++ b/src/remap/function.rs
@@ -22,7 +22,7 @@ mod md5;
 mod merge;
 mod now;
 mod only_fields;
-mod parse_aws_elb;
+pub mod parse_aws_elb;
 mod parse_duration;
 mod parse_grok;
 mod parse_json;
diff --git a/src/remap/function/parse_aws_elb.rs b/src/remap/function/parse_aws_elb.rs
index 8107d4eb0..056131d51 100644
--- a/src/remap/function/parse_aws_elb.rs
+++ b/src/remap/function/parse_aws_elb.rs
@@ -59,7 +59,7 @@ impl Expression for ParseAwsElbFn {
     }
 }
 
-fn parse_log(mut input: &str) -> Result<Value> {
+pub fn parse_log(mut input: &str) -> Result<Value> {
     let mut log = BTreeMap::new();
 
     macro_rules! get_value {
@@ -189,6 +189,88 @@ fn take_list(cond: impl Fn(char) -> bool) -> impl FnOnce(&str) -> SResult<Vec<&s
     }
 }
 
+pub fn parse_log_raw(mut input: &str) {
+    macro_rules! get_value {
+        ($name:expr, $parser:expr) => {{
+            let result: IResult<&str, _, (&str, nom::error::ErrorKind)> = $parser(input);
+            match result {
+                Ok((rest, value)) => {
+                    input = rest;
+                    value
+                }
+                Err(_) => return,
+            }
+        }};
+    }
+    macro_rules! field_raw {
+        ($name:expr, $parser:expr) => {
+            get_value!($name, $parser)
+        };
+    }
+    macro_rules! field {
+        ($name:expr, $($pattern:pat)|+) => {
+            field_raw!($name, preceded(char(' '), take_while1(|c| matches!(c, $($pattern)|+))))
+        };
+    }
+
+    field_raw!("type", take_while1(|c| matches!(c, 'a'..='z' | '0'..='9')));
+    field!("timestamp", '0'..='9' | '.' | '-' | ':' | 'T' | 'Z');
+    field_raw!("elb", take_anything);
+    field!("client_host", '0'..='9' | '.' | ':' | '-');
+    field!("target_host", '0'..='9' | '.' | ':' | '-');
+    field!("request_processing_time", '0'..='9' | '.' | '-');
+    field!("target_processing_time", '0'..='9' | '.' | '-');
+    field!("response_processing_time", '0'..='9' | '.' | '-');
+    field!("elb_status_code", '0'..='9' | '-');
+    field!("target_status_code", '0'..='9' | '-');
+    field!("received_bytes", '0'..='9' | '-');
+    field!("sent_bytes", '0'..='9' | '-');
+    let request = get_value!("request", take_quoted1);
+    let mut iter = request.splitn(2, ' ');
+    iter.next().unwrap();
+    match iter.next() {
+        Some(value) => {
+            let mut iter = value.rsplitn(2, ' ');
+            iter.next().unwrap();
+            iter.next().unwrap();
+        }
+        None => return,
+    };
+    field_raw!("user_agent", take_quoted2);
+    field_raw!("ssl_cipher", take_anything);
+    field_raw!("ssl_protocol", take_anything);
+    field_raw!("target_group_arn", take_anything);
+    field_raw!("trace_id", take_quoted2);
+    field_raw!("domain_name", take_quoted2);
+    field_raw!("chosen_cert_arn", take_quoted2);
+    field!("matched_rule_priority", '0'..='9' | '-');
+    field!(
+        "request_creation_time",
+        '0'..='9' | '.' | '-' | ':' | 'T' | 'Z'
+    );
+    field_raw!("actions_executed", take_quoted2);
+    field_raw!("redirect_url", take_quoted2);
+    field_raw!("error_reason", take_quoted2);
+}
+
+fn take_quoted2(input: &str) -> SResult<()> {
+    delimited(tag(" \""), until_quote2, char('"'))(input)
+}
+
+fn until_quote2(input: &str) -> SResult<()> {
+    let mut skip_delimiter = false;
+    for (i, ch) in input.char_indices() {
+        if ch == '\\' && !skip_delimiter {
+            skip_delimiter = true;
+        } else if ch == '"' && !skip_delimiter {
+            return Ok((&input[i..], ()));
+        } else {
+            skip_delimiter = false;
+        }
+    }
+    Err(nom::Err::Incomplete(nom::Needed::Unknown))
+}
+
 #[cfg(test)]
 mod tests {
     use super::*;
diff --git a/src/remap/mod.rs b/src/remap/mod.rs
index 43fb90cf0..fd9addd9a 100644
--- a/src/remap/mod.rs
+++ b/src/remap/mod.rs
@@ -1,4 +1,4 @@
-pub(crate) mod function;
+pub mod function;
 
 pub use function::*;
 use lazy_static::lazy_static;
```
</details>

Results:
```
parse_aws_elb/nom       time:   [43.395 us 43.907 us 44.529 us]
parse_aws_elb/nom_raw   time:   [7.4265 us 7.5076 us 7.6099 us]
parse_aws_elb/regex     time:   [50.973 us 51.548 us 52.259 us]
```

`parse_aws_elb/nom` return `remap::Value` with allocated data
`parse_aws_elb/nom_raw` only consume same data as `regex`
`parse_aws_elb/regex` is regex proposed in #5365.

Data allocation should take the nearly same time for both `nom`/`regex` but parsing looks like ~7x faster.

Some fields can have "no value" as `"-"`, should I use `Value::Null` in this case?

### CODE REVIEW

File: src/remap/function/parse_aws_elb.rs [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+        value_optional {
+            expr: |_| ParseAwsElbFn { value: Box::new(Noop) },
+            def: TypeDef { fallible: true, kind: value::Kind::Map },
+        }
+    ];
+
+    #[test]
+    fn parse_aws_elb() {
+        let logs = vec![
+            r#"http 2018-07-02T22:23:00.186641Z app/my-loadbalancer/50dc6c495c0c9188
```
Comment (jszwedko) [inline src/remap/function/parse_aws_elb.rs:211]: :+1: thanks for including all of their examples.

File: docs/reference/remap/parse_aws_elb.cue [hunk 2]
```diff
@@ -0,0 +1,76 @@
+package metadata
+
+remap: functions: parse_aws_elb: {
+	arguments: [
+		{
+			name:        "value"
+			description: "Access log of the Application Load Balancer."
```
Comment (jszwedko) [inline docs/reference/remap/parse_aws_elb.cue:7]: I think we may want to call this `parse_aws_alb` (or even `parse_aws_alb_log`) instead then to leave room for parsing classic load balancer (ELB) logs (https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/access-log-collection.html) and network load balancer (NLB) logs (https://docs.aws.amazon.com/elasticloadbalancing/latest/network/load-balancer-access-logs.html) in the future.

**Review by jszwedko — COMMENTED** [review]
Nice work! I appreciate the benchmarking to motivate the use of nom here.

I think mapping `-` to `Null` would be appropriate.

**Review by FungusHumungus — APPROVED** [review]
Nice.

**Review by jszwedko — APPROVED** [review]
Nice work!

**Pull request merged**